tools/check_copyright.sh

```diff
@@ -0,0 +1,99 @@
+#!/bin/bash
+
+# Copyright 2023, The OpenTelemetry Authors
+# SPDX-License-Identifier: Apache-2.0
+
+if [[ ! -e tools/check_copyright.sh ]]; then
+  echo "This tool must be run from the topmost directory." >&2
+  exit 1
+fi
+
+set -e
+
+#
+# Process input file .copyright-ignore,
+# - remove comments
+# - remove blank lines
+# to create file /tmp/all_ignored
+#
+
+grep -v "^#" < .copyright-ignore | \
+grep -v "^[[:space:]]*$" > /tmp/all_ignored
+
+#
+# Find all files from the repository
+# to create file /tmp/all_checked
+#
+
+find . -type f -print | sort -u > /tmp/all_checked
+
+#
+# Filter out /tmp/all_checked,
+# remove all ignored patterns from /tmp/all_ignored
+# When the pattern is *.md,
+# make sure to filter *\.md to avoid hiding *.cmd
+# Then, *\.md needs to be escaped to *\\.md,
+# to be given to egrep, hence the sed.
+# 
+
+while IFS= read -r PATTERN; do
+  SAFE_PATTERN=`echo "${PATTERN}" | sed "s!\.!\\\\\.!g"`
+  echo "Filtering out ${SAFE_PATTERN}"
+  egrep -v "${SAFE_PATTERN}" < /tmp/all_checked > /tmp/all_checked-tmp
+  mv /tmp/all_checked-tmp /tmp/all_checked
+done < /tmp/all_ignored
+
+#
+# For all files in /tmp/all_checked
+# - verify there is copyright
+# - verify there is a license
+# and append to /tmp/all_missing
+#
+# Valid copyright strings are:
+# - Copyright The OpenTelemetry Authors
+# - Copyright YEAR, The OpenTelemetry Authors
+# - Copyright YEAR, YEAR, The OpenTelemetry Authors
```

Comment:
Is the copyright string with YEAR special case for trace_id_ratio.cc, or there are other files too ?

Reply:
As usual with these cases, upfront warning : I am not a lawyer, nor do I represent one, etc.

Historically, the copyright in opentelemetry-cpp uses the first form, without year, for the most part.

In my experience however, I have always seen copyright applied with either:

* the year it is claimed (second form)
* the first year it is claimed, followed by the most recent year it is updated (third form).

For example, code created in 2020 will say "2020".

When later changed in 2021, it will say "2020, 2021".

When later changed in 2022, it will say "2020, 2022", and not use "2020, 2021, 2022".

Note that the Apache V2 license itself reads:

```
   APPENDIX: How to apply the Apache License to your work.

      To apply the Apache License to your work, attach the following
      boilerplate notice, with the fields enclosed by brackets "[]"
      replaced with your own identifying information. (Don't include
      the brackets!)  The text should be enclosed in the appropriate
      comment syntax for the file format. We also recommend that a
      file or class name and description of purpose be included on the
      same "printed page" as the copyright notice for easier
      identification within third-party archives.

   Copyright [yyyy] [name of copyright owner]
```

So, we are supposed to add a YEAR, for example:

`Copyright 2023, The OpenTelemetry Authors`

About file trace_id_ratio.cc, it contains copyright to OpenCensus as well as opentelemetry-cpp, so I wanted to preserve the dates for extra clarity.

My preference would be to use forms "YEAR" and "YEAR, YEAR" form everywhere, it is not specific to a file.

Reply:
Decision from today's SIG C++ meeting is to keep only `Copyright The OpenTelemetry Authors`.

Will adjust the patch accordingly.

Reply:
Fixed